### PR TITLE
Video Swin model added to kerashub

### PR DESCRIPTION
**Pull request by kernel-loophole**
- Added Video Swin backbone https://github.com/keras-team/keras-cv/tree/master/keras_cv/src/models/backbones/video_swin to kerasnlp models.
- Adapted configurations to allow multiple model versions under a single model with configurations handled via function arguments.
- Utilized KerasNLP's test routines for model verification.

https://github.com/keras-team/keras-hub/issues/1755
@divyashreepathihalli

### CODE REVIEW

**Comment by kernel-loophole** [comment]
Hi @divyashreepathihalli! I need some guidance on how to test this locally. Could you help me with the steps?

**Comment by kernel-loophole** [comment]
@divyashreepathihalli can you run test again.

**Review by mattdangerw — COMMENTED** [review]
Higher level feedback for a start.

- Remove all "aliases," that's a CV pattern we did not continue. All arch configurations are stored in json files we upload to Kaggle/HF. The "aliases" is just the preset name you want to use.
- Follow the task setup we have in KerasHub. I am guessing for this model we want to add a `VideoClassifier` task that we can model heavily on `ImageClassifier`.
- Remove all instances of `keras_cv` whereever they are.
- Remove the presets stuff until presets are actually upload.

This will need a much more substantial rewrite to match the abstractions of KerasHub, rather than a more direct copy.

**Comment by divyashreepathihalli** [comment]
@kernel-loophole are you still working on this? Please let us know once you are done addressing the comments.

**Comment by kernel-loophole** [comment]
@divyashreepathihalli yes ,i was bit busy ,will try to add this weekend

**Comment by kernel-loophole** [comment]
@mattdangerw update all cv pattern .configuration are in json file .it would be great if you can review and let me know your feedback on this .if you can provide any example on task setup that would also be great  .

File: keras_hub/src/models/video_swin/video_swin_aliases.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
+# Copyright 2024 The Kerashub Authors
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_aliases.py:1]: Please delete this file

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+@keras_hub_export("keras_hub_export.models.VideoSwinBackbone", package="keras_hub_export.models")
+class VideoSwinBackbone(Backbone):
+    """A Video Swin Transformer backbone model.
+
+    Args:
+        input_shape (tuple[int], optional): The size of the input video in
+            `(depth, height, width, channel)` format.
+            Defaults to `(32, 224, 224, 3)`.
+        input_tensor (KerasTensor, optional): Output of
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_backbone.py:51]: remove input_tensor

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            `keras.layers.Input()`) to use as video input for the model.
+            Defaults to `None`.
+        include_rescaling (bool, optional): Whether to rescale the inputs. If
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_backbone.py:54]: type hints are specified in this format
`arg_name: type. short description`
please refer other model implementation -ex:  https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/models/vit/vit_backbone.py

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+            set to `True`, inputs will be passed through a `Rescaling(1/255.0)` layer
+            and normalize with mean=[0.485, 0.456, 0.406] and std=[0.229, 0.224, 0.225].
+            Defaults to `False`.
+        patch_size (int | tuple(int)): The patch size for depth, height, and width
+            dimensions respectively. Default: (2,4,4).
+        embed_dim (int): Number of linear projection output channels.
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_backbone.py:60]: embedding_dim

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+            Default to 96.
+        depths (tuple[int]): Depths of each Swin Transformer stage.
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_backbone.py:62]: nit depths->depth

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+            Default to [3, 6, 12, 24]
+        window_size (int): The window size for depth, height, and width
+            dimensions respectively. Default to [8, 7, 7].
+        mlp_ratio (float): Ratio of mlp hidden dim to embedding dim.
+            Default to 4.
+        qkv_bias (bool): If True, add a learnable bias to query, key, value.
+            Default to True.
+        qk_scale (float): Override default qk scale of head_dim ** -0.5 if set.
+            Default to None.
+        drop_rate (float): Float between 0 and 1. Fraction of the input units to drop.
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_backbone.py:74]: dropout_rate

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+            Default: 0.2.
+        patch_norm (bool): If True, add layer normalization after patch embedding.
+            Default to False.
+
+    Example:
+    ```python
+    # Build video swin backbone without top layer
+    model = VideoSwinSBackbone(
+        include_rescaling=True, input_shape=(8, 256, 256, 3),
+    )
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_backbone.py:88]: add example with from_preset

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+    videos = keras.ops.ones((1, 8, 256, 256, 3))
+    outputs = model.predict(videos)
+    ```
+
+    References:
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_backbone.py:93]: mereg references in docstring description

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        - [Video Swin Transformer](https://arxiv.org/abs/2106.13230)
+        - [Official Code](https://github.com/SwinTransformer/Video-Swin-Transformer)
+    """  # noqa: E501
+
+    def __init__(
+        self,
+        *,
+        include_rescaling=False,
+        input_shape=(32, 224, 224, 3),
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_backbone.py:102]: any arg that changes the model architecture should not have default value

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+
+    References:
+        - [Video Swin Transformer](https://arxiv.org/abs/2106.13230)
+        - [Official Code](https://github.com/SwinTransformer/Video-Swin-Transformer)
+    """  # noqa: E501
+
+    def __init__(
+        self,
+        *,
+        include_rescaling=False,
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_backbone.py:101]: remove include_rescaling
this is moved into image converter

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
Hi @kernel-loophole the PR would need more updates to match KerasHub model implementation style. please follow this folder here as an example - https://github.com/keras-team/keras-hub/tree/master/keras_hub/src/models/sam
- remove aliases
- some args like rescaling, input_tensor etc have been dropped 
- follow docstring formats
- consistent arg names across similar model implementations.
- use standard test routines like in example folder - https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/models/sam/sam_backbone_test.py

**Comment by kernel-loophole** [comment]
@divyashreepathihalli tanks for review .will update that shortly .

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+            or input_spec.shape[-2] is None
+        ):
+            raise ValueError(
+                "Depth, height and width of the video must be specified"
+                " in `input_shape`."
+            )
+
+        x = input_spec
+
+        # if include_rescaling:
```
Comment (kernel-loophole) [inline keras_hub/src/models/video_swin/video_swin_backbone.py:132]: by changing the default value of scaling can change the model behavior

File: keras_hub/src/models/video_swin/video_swin_backbone.py (hunk 3 above)
Comment (divyashreepathihalli) [reply keras_hub/src/models/video_swin/video_swin_backbone.py]: yes plz

**Comment by divyashreepathihalli** [comment]
@kernel-loophole can you please sign the CLA? That would help trigger the CPU tests!

File: keras_hub/src/models/video_swin/video_swin_layers_test.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    VideoSwinPatchMerging,
+)
+from keras_hub.src.models.video_swin.video_swin_layers import (
+    VideoSwinWindowAttention,
+)
+from keras_hub.src.tests.test_case import TestCase
+
+
+class TestVideoSwinPatchingAndEmbedding(TestCase):
+    def test_patch_embedding_compute_output_shape(self):
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/video_swin/video_swin_layers_test.py:29]: use standerdized test routines for layers - refer to other models and layers

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
Hi @kernel-loophole I did a initial pass - did not look into the layer implementations much. They need to be implemented as layers and not models.
In general please refer to other model implementations to get an idea of design patterns and test routines. And implementation style comments are applicable throughout the code.

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
+# Copyright 2024 The Kerashub Authors
```
[the remark this answers is not in the document]
Comment (kernel-loophole) [reply keras_hub/src/models/video_swin/video_swin_backbone.py:1]: okay @divyashreepathihalli will update shortly .do i have to add an example of usage of this model

**Review by kernel-loophole — COMMENTED** [review]
update the changes

File: keras_hub/src/models/video_swin/video_swin_backbone.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+        #     # And use mean=[123.675, 116.28, 103.53] and
+        #     # std=[58.395, 57.12, 57.375] for normalization.
+        #     # So, if include_rescaling is set to True, then, to match with the
+        #     # official scores, following normalization should be added.
+        #     x = layers.Normalization(
+        #         mean=[0.485, 0.456, 0.406],
+        #         variance=[0.229**2, 0.224**2, 0.225**2],
+        #     )(x)
+
+        norm_layer = partial(layers.LayerNormalization, epsilon=1e-05)
```
[the remark this answers is not in the document]
Comment (kernel-loophole) [reply keras_hub/src/models/video_swin/video_swin_backbone.py:99]: anything else i can use

**Comment by kernel-loophole** [comment]
@divyashreepathihalli updated most of the it ,need help on some ,have a look ,you can also review to each commit also.

**Comment by sachinprasadhs** [comment]
@kernel-loophole , There are many failing test cases, could you please rebase the code and make the necessary changes. Thanks!

**Comment by kernel-loophole** [comment]
@sachinprasadhs sure will refactor this

**Comment by kernel-loophole** [comment]
@sachinprasadhs ready for review .

File: keras_hub/src/models/video_swin/video_swin_layers.py [hunk 15]
```diff
@@ -0,0 +1,960 @@
...
+
+    def __init__(
+        self, patch_size=(2, 4, 4), embed_dim=96, norm_layer=None, **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.patch_size = patch_size
+        self.embed_dim = embed_dim
+        self.norm_layer = norm_layer
+
+    def __compute_padding(self, dim, patch_size):
```
Comment (innat) [inline keras_hub/src/models/video_swin/video_swin_layers.py]: Use one underscore for private method.

File: keras_hub/src/models/video_swin/video_swin_layers.py [hunk 16]
```diff
@@ -0,0 +1,960 @@
...
+                "output_dim": self.output_dim,
+                "hidden_dim": self.hidden_dim,
+                "drop_rate": self.drop_rate,
+                "activation": self._activation_identifier,
+            }
+        )
+        return config
+
+
+class VideoSwinPatchingAndEmbedding(keras.layers.Layer):
```
Comment (innat) [inline keras_hub/src/models/video_swin/video_swin_layers.py:259]: This layer may need to implement `compute_output_shape` for jax-tpu training. 

```python
def compute_output_shape(self, input_shape):
        batch_size = input_shape[0]
        spatial_shape = input_shape[1:-1]

        padded = [
            dim + self._compute_padding(dim, self.patch_size[i])[1]
            for i, dim in enumerate(spatial_shape)
        ]
        out_spatial = [padded[i] // self.patch_size[i] for i in range(self.spatial_dims)]
        return (batch_size, *out_spatial, self.embed_dim)
```

File: keras_hub/src/models/video_swin/video_swin_layers.py [hunk 17]
```diff
@@ -0,0 +1,960 @@
...
+            self.input_dim * 3, use_bias=self.qkv_bias
+        )
+        self.attn_drop = keras.layers.Dropout(self.attn_drop_rate)
+        self.proj = keras.layers.Dense(self.input_dim)
+        self.proj_drop = keras.layers.Dropout(self.proj_drop_rate)
+        self.qkv.build(input_shape)
+        self.proj.build(input_shape)
+        self.built = True
+
+    def call(self, x, mask=None, training=None):
```
Comment (innat) [inline keras_hub/src/models/video_swin/video_swin_layers.py:509]: I'm not fully sure, but in keras 3.12, I got a warning

```
Layer 'x' (of type x) was passed an input with a mask attached to it.
However, this layer does not support masking and will therefore destroy the mask information.
Downstream layers will not see the mask.
```

It didn't happen with 3.8 or 3.9. To get rid of the warning, following flag needs to be added in the contructor

```python
def __init__(self, **kwargs):
    super().__init__(**kwargs)
    self.supports_masking = True
```

cc. @mattdangerw

File: keras_hub/src/models/video_swin/video_swin_layers.py [hunk 18]
```diff
@@ -0,0 +1,960 @@
...
+        self.depth = depth
+        self.qkv_bias = qkv_bias
+        self.qk_scale = qk_scale
+        self.drop_rate = drop_rate
+        self.attn_drop_rate = attn_drop_rate
+        self.drop_path_rate = drop_path_rate
+        self.norm_layer = norm_layer
+        self.downsampling_layer = downsampling_layer
+
+    def __compute_dim_padded(self, input_dim, window_dim_size):
```
Comment (innat) [inline keras_hub/src/models/video_swin/video_swin_layers.py]: one underscore

File: keras_hub/src/models/video_swin/video_swin_layers.py [hunk 19]
```diff
@@ -0,0 +1,960 @@
...
+        x = self.mlp(x)
+        x = self.drop_path(x, training=training)
+        return x
+
+    def call(self, x, mask_matrix=None, training=None):
+        shortcut = x
+        x = self.first_forward(x, mask_matrix, training)
+        x = shortcut + self.drop_path(x)
+        x = x + self.second_forward(x, training)
+        return x
```
Comment (innat) [inline keras_hub/src/models/video_swin/video_swin_layers.py:959]: This code pattern was taken directly from the official code. This is okay. But it can be modified a bit for better readability and official arch design. Some [pointer](https://github.com/innat/medic-ai/blob/5677e9ca4142317bac7d0130ef97de65b293b59a/medicai/models/swin/swin_layers.py#L686-L691).

File: keras_hub/src/models/video_swin/video_swin_layers.py [hunk 20]
```diff
@@ -0,0 +1,960 @@
...
+                    self.shift_size[2],
+                ),
+                axis=(1, 2, 3),
+            )
+        else:
+            x = shifted_x
+
+        # pad if required
+        if self.apply_pad:
+            return x[:, :depth, :height, :width, :]
```
Comment (innat) [inline keras_hub/src/models/video_swin/video_swin_layers.py]: This slicing could be a problem in jax-tpu cases. It might be safe to use [cropping layer](https://github.com/innat/medic-ai/blob/5677e9ca4142317bac7d0130ef97de65b293b59a/medicai/models/swin/swin_layers.py#L676).

File: keras_hub/src/models/video_swin/video_swin_layers.py (hunk 15 above)
Comment (kernel-loophole) [reply keras_hub/src/models/video_swin/video_swin_layers.py]: okay got it

File: keras_hub/src/models/video_swin/video_swin_layers.py (hunk 19 above)
Comment (kernel-loophole) [reply keras_hub/src/models/video_swin/video_swin_layers.py:959]: updated

**Comment by kernel-loophole** [comment]
@innat PR is ready for review.

**Comment by innat** [comment]
@kernel-loophole 
I can't review other than giving comment. 

The code is overall well-structured. There may (or may not) be a need for minor adjustments to fully align with the Keras Hub standards, which the Keras team is likely already aware of.

Additonally, you can createa collab gist to show how one can run this model. If needed, you can check [this](https://github.com/innat/VideoSwin/blob/main/guides/fine_tune/tf_videoswin_video_classification.ipynb) notebook. It uses sub-set of kinetics-400 dataset, small size dataset for quick prototyping. Also, if keras-team needs to verify the weight file, check [this](https://github.com/innat/VideoSwin/tree/main/guides/logit_checking) also. FYI, to run the model with full kinetrics-400 test set, I created a inference pipeline with full torch-style, [LINK](https://www.kaggle.com/code/ipythonx/kinetics-400-keras-pytorch-video-swin-evaluation) - matching the official code.

**Comment by kernel-loophole** [comment]
thanks really helpful .
